lib/module_player.js

```diff
@@ -0,0 +1,213 @@
+    // - If playModule is called more than once while goToModule is happening,
+    //   only the most recent becomes the nextModule.
+    // You only need to wait for playModule if you want to be sure that the
+    // player has tried to transition to your module, which is perfect for
+    // tests.
+    async playModule(module) {
+      this.nextModule = module;
+      if (monitor.isEnabled()) {
+        monitor.update({
+          event: `playModule: ${module.name}`,
+          time: time.now(),
+          deadline: module.deadline
+        });
+      }
+      await this.performTransitions();
+    }
+
+    async performTransitions() {
+      if (this.transitionInProgress) {
+        return; // We'll do this work on the next iteration of the loop below.
+      }
+      this.transitionInProgress = true;
+      while (this.oldModule != this.nextModule) {
+        const module = this.nextModule;
+        // Our local copy of the module we are trying to go to.
+        await this.goToModule(this.nextModule);
+        // Maybe we succeeded in going to module... maybe we didn't. It isn't
+        // clear. If we failed to go to the module, then we should stop trying
+        // to do so, but only if we haven't been told to go somewhere else.
+        const failedToTransition = module != this.oldModule;
+        const stillTryingToGetThere = module == this.nextModule;
+        if (failedToTransition && stillTryingToGetThere) {
+          // Just stop the loop, then.
+          this.nextModule = this.oldModule;
+        }
+      }
+      this.transitionInProgress = false;
+    }
+    /**
+     * Goes to a specified module using a transition. This will take some time to
+     * happen. We aren't exactly how much time it will take, but we know that
+     * we'll endeavor to begin the visible transition by the module's deadline.
+     *
+     * This method can throw if something went wrong in preparing the module for
+     * display. This method also returns false to indicate that we should skip
+     * the usual delay until the next module and switch asap. Once a transition
+     * starts, there's no way to interrupt it, so we change to the next module
+     * as soon as we can.
+     */
+    async goToModule(module) {
+      // Instantiate the module. If this throws, by contract, we have no cleanup
+      // to do, so allow the exception to bubble up to the error logger and await
+      // being told which module to go to next.
+      try {
+        debug(`Instantiating module ${module.name}.`);
+        await Promise.race([module.instantiate(), delayThenReject(5000)]);
+      } catch (e) {
+        // If this throws, it's because the delayed rejection.
+        debug(`Module ${module.name} timed out in instantiation.`);
+        module.dispose();
+        return;
+      }
+      // Instantiation is complete. If while we were doing that, we were told to
+      // go to a new module, abort!
```

Comment:
This all feels like it's trying to hard to only instantiate one module at a time. So what if we instantiate a module1 and, moments later, decide that no no we want module2, and instantiate module2 as well. Would the code then be cleaner if there were an async chain like "instantiate this module, then if it's still the intended target prepare it to be shown, then if it's still the target begin transitioning to it, and on reject dispose it"? It's the same logic, it just feels like this is a very big monolithic code path that might read better as chained ops. (or maybe it's late).

Reply:
I tried an approach where we had Commands and Steps, some of which were interruptible, etc etc, but the result wasn't particularly clean. I think this is easier to read.